Code/Framework/AzFramework/Tests/Spawnable/SpawnableScriptMediatorTests.cpp

```diff
@@ -0,0 +1,302 @@
+            specializations.Append("test");
+            specializations.Append("spawnable");
+        }
+    };
+    
+    class SpawnableScriptMediatorTests
+        : public AllocatorsFixture
+        , public AzFramework::Scripts::SpawnableScriptNotificationsBus::MultiHandler
+    {
+    public:
+        using TicketToEntitiesPair = AZStd::pair<AzFramework::EntitySpawnTicket, AZStd::vector<AZ::EntityId>>;
+        constexpr static AZ::u64 EntityIdStartId = 40;
+
+        AZStd::vector<TicketToEntitiesPair> m_spawnedEntities;
+
+        void OnSpawn(AzFramework::EntitySpawnTicket spawnTicket, AZStd::vector<AZ::EntityId> entityList) override
+        {
+            m_spawnedEntities.push_back({ spawnTicket, AZStd::move(entityList) });
+        }
+
+        virtual void OnDespawn(AzFramework::EntitySpawnTicket spawnTicket) override
+        {
+            AZStd::erase_if(
+                m_spawnedEntities,
+                [&spawnTicket](const TicketToEntitiesPair& pair) -> bool
+                {
+                    return pair.first == spawnTicket;
+                });
+        }
+
+        void SetUp() override
+        {
+            AllocatorsFixture::SetUp();
+
+            m_application = new TestApplication();
+            AZ::ComponentApplication::Descriptor descriptor;
+            m_application->Start(descriptor);
+
+            // Without this, the user settings component would attempt to save on finalize/shutdown. Since the file is
+            // shared across the whole engine, if multiple tests are run in parallel, the saving could cause a crash
+            // in the unit tests.
+            AZ::UserSettingsComponentRequestBus::Broadcast(&AZ::UserSettingsComponentRequests::DisableSaveOnFinalize);
+
+            auto managerInterface = AzFramework::SpawnableEntitiesInterface::Get();
+            m_manager = azrtti_cast<AzFramework::SpawnableEntitiesManager*>(managerInterface);
+        }
+
+        void TearDown() override
+        {
+            AzFramework::Scripts::SpawnableScriptNotificationsBus::MultiHandler::BusDisconnect();
+            // One more tick on the spawnable entities manager in order to delete the ticket fully.
+            while (m_manager->ProcessQueue(
+                       AzFramework::SpawnableEntitiesManager::CommandQueuePriority::High |
+                       AzFramework::SpawnableEntitiesManager::CommandQueuePriority::Regular) !=
+                   AzFramework::SpawnableEntitiesManager::CommandQueueStatus::NoCommandsLeft)
+                ;
+            
+            m_spawnedEntities = {};
+
+            delete m_application;
+            m_application = nullptr;
+
+            AllocatorsFixture::TearDown();
+        }
+
+        void WaitForResponse(AzFramework::Scripts::SpawnableScriptMediator& mediator)
+        {
+            for (size_t i=0; i<1000; ++i) // Don't do this indefinitely to avoid deadlocking on a failing test.
```

Comment:
The fact that it requires the queue to be processed a random number of times sounds prone to errors in the future. Since you are authoring the test, it should be possible to know how many times you need to process the queue as well right? Maybe we can change this function to be WaitForResponse(SpawnableScriptMediator, numberOfTimesToProcessQueue) and each individual test can specify how many times the queue needs to be processed? This way, the test can actually fail if the queue processing did not yield the required results.

I know that this has taken from SpawnableEntitiesManagerTests.cpp. So pinging @AMZN-koppersr to understand the original thought process behind this and whether specifying the number of process queue requests makes sense.

Reply:
The SpawnableEntitiesManager is a black box for these tests, it should be tested in SpawnableEntitiesManagerTests. While I could estimate that number of requests my operations consume, it's not guaranteed to stay consistent if the underlying code will be modified in the future. I think it may make sense to just keep the 1000 limit here. Though ideally there would be an API to ProcessAllCommands or something similar.

Reply:
discussed in slack

Reply:
fyi, this is the reasoning behind the modification
```
In certain scenarios, the SpawnableEntitiesManager requeues the commands while processing them
`return CommandResult::Requeue;`

When they get requeued, the queue needs to be processed again
. It happens in scenarios where the request ids don't match or if the spawnable is not ready.

 I believe requestids don't match when there are Barrier() and LoadBarrier() commands involved which pause processing of all further operations on a ticket until previous requests are completed
.
You won't hit any of these scenarios in the unit tests. That's why processing the queue once will work for you. I think it's better for the test framework to be in control of the number of times we need to process until we are truly asynchronous. We can change it if we run into issues with these in the future.
```